interfaces: new "ros" host file system support

MirkoFerrati opened the pull request:
This PR adds a new "ros" interface that allows snaps to access the host "/opt/ros" paths. This allows ros snaps to access to shared configuration and similar data.

MirkoFerrati commented:
@Meulengracht I don't know why the CLA check failed, I signed it again just in case. Also PR title updated.

pedronis commented on interfaces/builtin/ros_opt_data.go:
```diff
@@ -0,0 +1,65 @@
+const rosOptDataConnectedPlugAppArmor = `
+# Description: Allows read-only access to the xacro,yaml,urdf,stl files in /opt/ros folder
+# capability dac_read_search,  # this should not be necessary, the idea is to read only what normal user can read
+/var/lib/snapd/hostfs/opt/ros/ r,
```
is the fact that they are not at /opt/ros a big annoyance or not? I suppose not given you wrote this way, but wondering because in principle the interface could also maybe mount things there inside the snap namespace?

MirkoFerrati replied:
Indeed I also considered mounting them in the same /opt/ros convenient folder as the host but I felt it would be confusing for some kind of usages. 
For example if the snap using this interface is already putting something else in /opt/ros and would see something overwritten or shadowed. 
Another example is if the Ros version in the host is not fully compatible with the Ros version inside the snap -> the default folder would be loaded automatically with broken data, while a separate folder can be managed with additional checks. This is mostly true for *.so libraries which I am filtering anyway, but not only for libraries.

pedronis reviewed (commented):
this looks reasonable, a question

alexmurray commented on interfaces/builtin/ros_opt_data.go:
```diff
@@ -0,0 +1,65 @@
+# capability dac_read_search,  # this should not be necessary, the idea is to read only what normal user can read
+/var/lib/snapd/hostfs/opt/ros/ r,
+/var/lib/snapd/hostfs/opt/ros/**/ r,
+/var/lib/snapd/hostfs/opt/ros/**/*.config r,        # for robot configuration
```
why not just allow read only access to everything under `/opt/ros/`? why only restrict it to these particular file extensions?

MirkoFerrati replied:
Just to be clear from the beginning, I know filtering by extensions is not related to security since any file can be renamed in the host to match the filter and be accessible from the snap.
The reason I filtered is to clearly define what should be used and what not. In particular *.so files and executables are not going to be compatible by default unless the host uses the same Ros version as the snap. 
While we want this interface to be used for common file formats related to robotics data. Data files are usually compatible with multiple versions of Ros/Ubuntu and are not well suited to be distributed inside application snaps, so it's convenient to keep them in the host and use them in the snaps.

alexmurray replied:
Should the interface perhaps be called something like `ros-host-data` then instead?

MirkoFerrati replied:
@alexmurray  no strong opinion on naming, I think I followed an advice from a meeting with Michael Vogt but it was many months ago. 
Isn't "host" implied by the concept of interface itself (interfacing snaps with the host system)? 
With "opt" I am telling users they cannot put their ros data in a random place in their host and expect it to be visible inside the snap.

pedronis replied:
we never use "host" interface names, except hostname-control, usually in the few cases that we want to point out about the host we use "system" like in system-packages-doc,  if we follow that convention it would become "system-ros-opt-data" otherwise ros-opt-data is also fine I think

MirkoFerrati reviewed (commented):
Thanks @pedronis @alexmurray for the review, I answered in comments. 
do you see anything that I should be improving or any test that may not work?

MirkoFerrati commented:
@pedronis  how do I run a full CI run? it seems status is stuck on waiting to be reported

pedronis commented:
@MirkoFerrati 

> @pedronis how do I run a full CI run? it seems status is stuck on waiting to be reported

we changed how our CI suites were organized, you need to merge in master or rebase on it

MirkoFerrati commented:
> @MirkoFerrati
> 
> > @pedronis how do I run a full CI run? it seems status is stuck on waiting to be reported
> 
> we changed how our CI suites were organized, you need to merge in master or rebase on it

rebased, now it is appropriately waiting for approval @pedronis

pedronis commented on interfaces/builtin/ros_opt_data.go:
```diff
@@ -0,0 +1,65 @@
+    allow-installation:
+      slot-snap-type:
+        - core
+    deny-auto-connection: false
```
I missed this, we what's the intention here? that any snap can access this without asking for permission? in that case we usually omit this, if the intention is that snap ask for autoconnection then this should be true

MirkoFerrati replied:
I explicitly set the value as I did not know the usual behavior, but if the coding style is to omit if true and it goes true by default, I will remove this line.

pedronis replied:
sorry, I'm confused, you set it to false, so not sure why you mention true in your sentence. Should any snap installed on a system get access to these files? there no security consequences to that? I suppose so, but want to double check.

MirkoFerrati replied:
sorry, my intention is for snaps to get autoconnected automatically, so this should be false because this is **deny**-auto-connection. My sentence above was written too quickly, if the default behavior is to get autoconnected automatically == true  -> I can omit this and remove the line. Already pushed without it.

pedronis reviewed (commented):
comment and a question about something I didn't spot the first time looking at this

pedronis commented on tests/main/interfaces-ros-opt-data/test-snapd-sh/meta/snap.yaml:
```diff
@@ -0,0 +1,13 @@
+plugs:
+    opt-ros:  
+        interface: ros-opt-data
+        read: [/opt/ros]
```
this attribute is irrelevant no?

MirkoFerrati replied:
correct, I removed it. On this regard, I would like to cleanup history or squash all commits together, what is the policy? Will you squash when merging? Should I wait for final review before force-pushing?

pedronis replied:
usually is best to force push only at the end and not during reviews, but yes we can squash merge this when we merge it, we have also a reminder label for that, I added it

pedronis reviewed (approved):
thanks, the test snap seems to use an irrelevant attribute

MirkoFerrati commented:
@alexmurray  did I address all your comments? do you need anything before approval?

pedronis commented on interfaces/builtin/ros_opt_data.go:
```diff
@@ -0,0 +1,64 @@
+	registerIface(&rosOptDataInterface{commonInterface{
+		name:                  "ros-opt-data",
+		summary:               rosOptDataSummary,
+		implicitOnCore:        true,
```
@MirkoFerrati how would this be used on Ubuntu Core though? who would put data into /opt/ros on Core?

MirkoFerrati replied:
Agreed, this is not a use-case for Core. It's for standard systems that want to start using some ROS applications from snaps while the rest stays in the host. 
If instead a system is running UC it means the developers are only using snaps and we have nice tutorials on how this kind of shared data can be stored inside snaps.
I pushed a commit setting this to False

alexmurray reviewed (approved):
LGTM except for the outstanding question regarding `implicitOnCore` - I also agree that this feels like it should be `false` rather than `true`.

pedronis commented:
blocked pending answer to question

MirkoFerrati reviewed (commented):
change done and commented @pedronis

Meulengracht commented:
@MirkoFerrati mind rebasing on top of master? Then i'll approve a test run

MirkoFerrati commented:
> @MirkoFerrati mind rebasing on top of master? Then i'll approve a test run

@Meulengracht there you go, force-pushed on top of current master

Meulengracht commented:
> > @MirkoFerrati mind rebasing on top of master? Then i'll approve a test run
> 
> @Meulengracht there you go, force-pushed on top of current master

The unit tests are failing due to the new interface. Please fix this
